truenas_api_client/__init__.py

```diff
@@ -314,85 +313,76 @@ def _send(self, data):
+        try:
+            if 'method' in message:
+                params = message['params']
+                match message['method']:
+                    case 'collection_update':
+                        if self._event_callbacks:
+                            if '*' in self._event_callbacks:
+                                for event in self._event_callbacks['*']:
+                                    self._run_callback(event, [params['msg'].upper()], params)
+                            if params['collection'] in self._event_callbacks:
+                                for event in self._event_callbacks[params['collection']]:
+                                    self._run_callback(event, [params['msg'].upper()], params)
+                    case 'notify_unsubscribed':
+                        if params['collection'] in self._event_callbacks:
+                            for event in self._event_callbacks[params['collection']]:
+                                if 'error' in params:
+                                    event['error'] = params['error']['reason'] or params['error']
+                                event['event'].set()
+                    case _:
+                        logger.error('Received unknown notification %r', message['method'])
+            elif 'id' in message:
+                if self._set_options_call and message['id'] == self._set_options_call.id:
                     if 'error' in message:
-                        event['error'] = message['error']['reason'] or message['error']['error']
-                    event['ready'].set()
-                    event['event'].set()
+                        try:
+                            self._parse_error(message['error'], self._set_options_call)
+                        except Exception:
+                            logger.error('Unhandled exception in Client._parse_error', exc_info=True)
+                        else:
+                            logger.error('Error setting client options: %r', self._set_options_call.error)
+                    self._connected.set()
+                elif call := self._calls.get(message['id']):
+                    if 'result' in message:
+                        call.result = message['result']
+                    if 'error' in message:
+                        try:
+                            self._parse_error(message['error'], call)
+                        except Exception:
+                            logger.error('Unhandled exception in Client._parse_error', exc_info=True)
+                    call.returned.set()
+                    self._unregister_call(call)
+                else:
+                    logger.error('Received a response for non-registered method call %r', message['id'])
+            else:
+                logger.error('Received unknown message %r', message)
+        except Exception:
+            logger.error('Unhandled exception in Client._recv', exc_info=True)
+
+    def _parse_error(self, error: dict, call: Call):
```

Comment:
Just an observation as I'm going through documenting this in [NAS-129755](https://github.com/truenas/api_client/tree/NAS-129755). `message['error']['message']` is being received as part of the JSONRPC v2 protocol, but it's not accessed anywhere.
Would it be better to replace usage of `message['error']['data']['reason']`, a nonstandard field, with `message['error']['message']`?

Reply:
`message['error']['data']['errname']` is also being received and never accessed.

Reply:
`message['error']['message']` is not accessed because we always have a better error message.

`message['error']['data']['errname']` is used in the UI.

Reply:
Would it be better to make use of the standard field `message['error']['message']` by storing `message['error']['data']['reason']` there instead and eliminating the field `message['error']['data']['reason']`?

Reply:
It's just a bit more standardized since `message['error']['message']` is defined in the JSON-RPC 2.0 specification whereas `message['error']['data']['reason']` is a custom field. Hardly makes a difference except that it's one less field we're sending in our messages.

Reply:
Specification (https://www.jsonrpc.org/specification#error_object) defines `message['error']['message']` as

> A String providing a short description of the error.
> The message SHOULD be limited to a concise single sentence.

`message['error']['data']['reason']` contains `CallError` text, subprocess stderr and other things, which can extend far beyond what can be described as a "concise single sentence"

Reply:
Yeah that's fair